`GetIssuer` just panics rather than issuing a useful error when there's no networking

**Description**

`GetIssuer` panics when there's no network access

https://github.com/sigstore/fulcio/blob/v0.2.0/pkg/config/config_test.go#L104

```
PASS
ok      github.com/sigstore/fulcio/pkg/challenges       0.270s
=== RUN   TestMetaURLs
=== RUN   TestMetaURLs/AWS_meta_URL
=== RUN   TestMetaURLs/GKE_meta_URL
--- PASS: TestMetaURLs (0.00s)
    --- PASS: TestMetaURLs/AWS_meta_URL (0.00s)
    --- PASS: TestMetaURLs/GKE_meta_URL (0.00s)
=== RUN   TestLoad
--- FAIL: TestLoad (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x753673]

goroutine 9 [running]:
testing.tRunner.func1.2({0x79cc00, 0xada640})
        testing/testing.go:1209 +0x24e
testing.tRunner.func1()
        testing/testing.go:1212 +0x218
panic({0x79cc00, 0xada640})
        runtime/panic.go:1038 +0x215
github.com/sigstore/fulcio/pkg/config.(*FulcioConfig).GetIssuer(0x0, {0x7fea68, 0x1b})
        github.com/sigstore/fulcio/pkg/config/config.go:85 +0x93
github.com/sigstore/fulcio/pkg/config.TestLoad(0xc0000b9860)
        github.com/sigstore/fulcio/pkg/config/config_test.go:104 +0x165
testing.tRunner(0xc0000b9860, 0x814b28)
        testing/testing.go:1259 +0x102
created by testing.(*T).Run
        testing/testing.go:1306 +0x35a
FAIL    github.com/sigstore/fulcio/pkg/config   0.006s
```

**To reproduce**

1. have the repo
2. run `go mod vendor` so we have the dependencies available
3. `docker run -v $PWD:/fulcio --network=none -it golang`
4. `cd /fulcio` and `go test ./... -v`

**Remediation**

There should probably be an error that's more useful than a panic

Also it'd be nice for all tests that require networking to be tagged so we can avoid them within our hermetic nix build environment that doesn't have network access :slightly_smiling_face:

--- a/pkg/config/config_network_test.go
+++ b/pkg/config/config_network_test.go
@@ -1,0 +1,92 @@
+// Copyright 2022 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//go:build !hermetic
+
+package config
+
+import (
+	"context"
+	"io/ioutil"
+	"path/filepath"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+)
+
+func TestLoad(t *testing.T) {
+	td := t.TempDir()
+	cfgPath := filepath.Join(td, "config.json")
+	if err := ioutil.WriteFile(cfgPath, []byte(validCfg), 0644); err != nil {
+		t.Fatal(err)
+	}
+
+	cfg, _ := Load(cfgPath)
+	got, ok := cfg.GetIssuer("https://accounts.google.com")
+	if !ok {
+		t.Error("expected true, got false")
+	}
+	if got.ClientID != "foo" {
+		t.Errorf("expected foo, got %s", got.ClientID)
+	}
+	if got.IssuerURL != "https://accounts.google.com" {
+		t.Errorf("expected https://accounts.google.com, got %s", got.IssuerURL)
+	}
+	if got := len(cfg.OIDCIssuers); got != 1 {
+		t.Errorf("expected 1 issuer, got %d", got)
+	}
+
+	got, ok = cfg.GetIssuer("https://oidc.eks.fantasy-land.amazonaws.com/id/CLUSTERIDENTIFIER")
+	if !ok {
+		t.Error("expected true, got false")
+	}
+	if got.ClientID != "bar" {
+		t.Errorf("expected bar, got %s", got.ClientID)
+	}
+	if got.IssuerURL != "https://oidc.eks.fantasy-land.amazonaws.com/id/CLUSTERIDENTIFIER" {
+		t.Errorf("expected https://oidc.eks.fantasy-land.amazonaws.com/id/CLUSTERIDENTIFIER, got %s", got.IssuerURL)
+	}
+
+	if _, ok := cfg.GetIssuer("not_an_issuer"); ok {
+		t.Error("no error returned from an unconfigured issuer")
+	}
+}
+
+func TestLoadDefaults(t *testing.T) {
+	td := t.TempDir()
+
+	// Don't put anything here!
+	cfgPath := filepath.Join(td, "config.json")
+	cfg, err := Load(cfgPath)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if diff := cmp.Diff(DefaultConfig, cfg, cmpopts.IgnoreUnexported(FulcioConfig{})); diff != "" {
+		t.Errorf("DefaultConfig(): -want +got: %s", diff)
+	}
+
+	ctx := context.Background()
+
+	if got := FromContext(ctx); nil != got {
+		t.Errorf("FromContext(): %#v, wanted nil", got)
+	}
+
+	ctx = With(ctx, cfg)
+	if diff := cmp.Diff(cfg, FromContext(ctx), cmpopts.IgnoreUnexported(FulcioConfig{})); diff != "" {
+		t.Errorf("FromContext(): -want +got: %s", diff)
+	}
+}

--- a/pkg/config/config_test.go
+++ b/pkg/config/config_test.go
@@ -16,14 +16,7 @@
 package config
 
 import (
-	"context"
-	"io/ioutil"
-	"os"
-	"path/filepath"
 	"testing"
-
-	"github.com/google/go-cmp/cmp"
-	"github.com/google/go-cmp/cmp/cmpopts"
 )
 
 var validCfg = `
@@ -41,13 +34,6 @@ var validCfg = `
 	}
 }
 `
-
-// Skip test when HERMETIC=true is set, since config tests require network access
-func skipHermetic(t *testing.T) {
-	if os.Getenv("HERMETIC") != "" {
-		t.Skip("Skipping testing in hermetic test environment")
-	}
-}
 
 func TestMetaURLs(t *testing.T) {
 	tests := []struct {
@@ -98,73 +84,5 @@ func TestMetaURLs(t *testing.T) {
 				}
 			}
 		})
-	}
-}
-
-func TestLoad(t *testing.T) {
-	skipHermetic(t)
-
-	td := t.TempDir()
-	cfgPath := filepath.Join(td, "config.json")
-	if err := ioutil.WriteFile(cfgPath, []byte(validCfg), 0644); err != nil {
-		t.Fatal(err)
-	}
-
-	cfg, _ := Load(cfgPath)
-	got, ok := cfg.GetIssuer("https://accounts.google.com")
-	if !ok {
-		t.Error("expected true, got false")
-	}
-	if got.ClientID != "foo" {
-		t.Errorf("expected foo, got %s", got.ClientID)
-	}
-	if got.IssuerURL != "https://accounts.google.com" {
-		t.Errorf("expected https://accounts.google.com, got %s", got.IssuerURL)
-	}
-	if got := len(cfg.OIDCIssuers); got != 1 {
-		t.Errorf("expected 1 issuer, got %d", got)
-	}
-
-	got, ok = cfg.GetIssuer("https://oidc.eks.fantasy-land.amazonaws.com/id/CLUSTERIDENTIFIER")
-	if !ok {
-		t.Error("expected true, got false")
-	}
-	if got.ClientID != "bar" {
-		t.Errorf("expected bar, got %s", got.ClientID)
-	}
-	if got.IssuerURL != "https://oidc.eks.fantasy-land.amazonaws.com/id/CLUSTERIDENTIFIER" {
-		t.Errorf("expected https://oidc.eks.fantasy-land.amazonaws.com/id/CLUSTERIDENTIFIER, got %s", got.IssuerURL)
-	}
-
-	if _, ok := cfg.GetIssuer("not_an_issuer"); ok {
-		t.Error("no error returned from an unconfigured issuer")
-	}
-}
-
-func TestLoadDefaults(t *testing.T) {
-	skipHermetic(t)
-
-	td := t.TempDir()
-
-	// Don't put anything here!
-	cfgPath := filepath.Join(td, "config.json")
-	cfg, err := Load(cfgPath)
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	if diff := cmp.Diff(DefaultConfig, cfg, cmpopts.IgnoreUnexported(FulcioConfig{})); diff != "" {
-		t.Errorf("DefaultConfig(): -want +got: %s", diff)
-	}
-
-	ctx := context.Background()
-
-	if got := FromContext(ctx); nil != got {
-		t.Errorf("FromContext(): %#v, wanted nil", got)
-	}
-
-	ctx = With(ctx, cfg)
-	if diff := cmp.Diff(cfg, FromContext(ctx), cmpopts.IgnoreUnexported(FulcioConfig{})); diff != "" {
-		t.Errorf("FromContext(): -want +got: %s", diff)
 	}
 }

--- a/pkg/config/config_test.go
+++ b/pkg/config/config_test.go
@@ -18,6 +18,7 @@ package config
 import (
 	"context"
 	"io/ioutil"
+	"os"
 	"path/filepath"
 	"testing"
 
@@ -40,6 +41,13 @@ var validCfg = `
 	}
 }
 `
+
+// Skip test when HERMETIC=true is set, since config tests require network access
+func skipHermetic(t *testing.T) {
+	if os.Getenv("HERMETIC") != "" {
+		t.Skip("Skipping testing in hermetic test environment")
+	}
+}
 
 func TestMetaURLs(t *testing.T) {
 	tests := []struct {
@@ -94,6 +102,8 @@ func TestMetaURLs(t *testing.T) {
 }
 
 func TestLoad(t *testing.T) {
+	skipHermetic(t)
+
 	td := t.TempDir()
 	cfgPath := filepath.Join(td, "config.json")
 	if err := ioutil.WriteFile(cfgPath, []byte(validCfg), 0644); err != nil {
@@ -132,6 +142,8 @@ func TestLoad(t *testing.T) {
 }
 
 func TestLoadDefaults(t *testing.T) {
+	skipHermetic(t)
+
 	td := t.TempDir()
 
 	// Don't put anything here!

